Migrate HTTP api

danmason opened the pull request:
Resolves #1032, #803  - using reitit for routing and proper organized coercion of params against a `spec`.

TODO:
- [x] Clarify the use of crux/ids in the main API - currently need write handlers to pass them over the wire, should this be something we are exposing anyway?
- [x] Merge regular + console endpoints, and under same handler
- [x] Use reitit for routing
- [x] Ensure all routes use muuntaja for negotiation (?)
--> This currently supports EDN and transit+json as outputs, only currently accept EDN inputs
- [x] Ensure all non console routes use clojure API rather than Java API
- [x] Use reitit for handling spec validation + coercion of params
- [x] Migrate remote API to use new routes
- [x] Error messages for spec/coercion fail.
- [x] Cleanup remote API - no string encoding.
- [x] Cleanup the http-server namespace of unused content.
- [x] All endpoints prefixed with `_crux`.
- [x] Should empty `attribute-stats` call return an empty map or a nil?
- [x] Update REST API docs
- [x] Remove HTTP Special case from thrown (API test)
- [x] Remote API use post `query` - ensure ui route test for GET query still works.
- [x] Add transit decoder for body - muuntaja (and test it)
- [ ] Exception handler for muuntaja decode - likely want to think further on exception handlers
- [ ] May want some more specific tests for exceptions in `ui-routes-test`

Among these changes is a refactor and cleanup to the behaviour of `http-server.entity`.

Note: Errors are now (generally) thrown across the board, and error handling is done before muuntaja formatting (such that errors will now be formatted by muuntaja) - this also means that coercion errors within `query`/`entity` will still result in the user getting a page back when requesting HTML.

Likely some updates wanted to the REST API changes (minor tweaks, and perhaps a small explainer on types you're expected to pass?) - can see it here https://www.opencrux.com/http/reference/http.html

danmason commented on crux-http-client/src/crux/remote_api_client.clj:
```diff
@@ -113,21 +113,35 @@
     valid-time (assoc :valid-time valid-time)
     transact-time (assoc :transact-time transact-time)))
 
+(defn edn->query-params
```
No need for this - can always use `q` param to pass the query.

danmason commented on crux-http-server/src/crux/http_server.clj:
```diff
@@ -1,269 +1,194 @@
-    :headers headers
-    :body body}))
 
 (defn- redirect-response [url]
```
Unnecessary - see ring util response

jarohen commented on crux-http-client/src/crux/remote_api_client.clj:
```diff
@@ -1,26 +1,23 @@
         open-paren \(]
-    (when-not (= (int open-paren) (.read in))
-      (throw (RuntimeException. "Expected delimiter: (")))
+    (let [first-char (.read in)]
```
reckon this'd be clearer as a `cond`, maybe even a `(condp = (.read in) ...)`

jarohen commented on crux-http-client/src/crux/remote_api_client.clj:
```diff
@@ -219,71 +221,75 @@
-                      {:http-opts {:method :get}
-                       :->jwt-token ->jwt-token}))
+    (->
+     (api-request-sync (str url "/_crux/await-tx-time" )
```
nit: trailing space, and can we have the first arg to `->` on the same line as the `->`?

jarohen commented on crux-http-client/src/crux/remote_api_client.clj:
```diff
@@ -219,71 +221,75 @@
-    (api-request-sync (cond-> (str url "/await-tx?tx-id=" (:crux.tx/tx-id tx))
-                        timeout (str "&timeout=" (cio/format-duration-millis timeout)))
-                      {:http-opts {:method :get}
+    (api-request-sync (str url "/_crux/await-tx" )
```
nit: another cheeky trailing space

jarohen commented on crux-http-server/src-cljs/crux/ui/events.cljs:
```diff
@@ -20,6 +20,7 @@
                               (str "meta[title=" title "]"))
                              (.getAttribute "content"))
          edn-content (reader/read-string {:readers {'object pr-str
+                                                    'crux/id #(str %)
```
nit: `#(str %)` === `str`

jarohen commented on crux-http-server/src-cljs/crux/ui/http.cljs:
```diff
@@ -12,7 +12,10 @@
-                                (edn/read-string {:readers {'crux.http/entity-ref entity-ref/->EntityRef}} (-body xhrio)))
+                                (edn/read-string
+                                 {:readers {'crux.http/entity-ref entity-ref/->EntityRef
+                                            'crux/id #(str %)}}
```
nit: `#(str %)` === `str`

jarohen commented on crux-http-server/src/crux/http_server/entity.clj:
```diff
@@ -128,61 +149,67 @@
 (defn ->entity-html-encoder [opts]
   (reify mfc/EncodeToBytes
-    (encode-to-bytes [_ {:keys [eid no-entity? not-found? error entity ^Closeable entity-history] :as res} charset]
+    (encode-to-bytes [_ {:keys [eid no-entity? not-found? cause entity ^Closeable entity-history] :as res} charset]
```
why `cause` here rather than `error`?

jarohen commented on crux-http-server/src/crux/http_server/entity.clj:
```diff
@@ -128,61 +149,67 @@
+                             (finally
+                               (cio/try-close entity-history)))
+             entity (print-method entity w)
+            :else (.write w ^String (pr-str res))))))))
```
this seems like it'd return something like `{:entity nil}` for a not-found response? think we said to leave the response body empty in this case?

jarohen commented on crux-http-server/src/crux/http_server/entity.clj:
```diff
@@ -128,61 +149,67 @@
+          (cond
+            entity-history (try
+                             (if (.hasNext entity-history)
+                               (print-method (iterator-seq entity-history) w)
```
might be clearer to go with `(print-method (or (iterator-seq entity-history) '()) w) ` rather than this `if`?

jarohen commented on crux-http-server/src/crux/http_server/entity.clj:
```diff
@@ -128,61 +149,67 @@
+            entity (transit/write w entity)
+            entity-history (try
+                             (if (.hasNext entity-history)
+                               (transit/write w (iterator-seq entity-history))
```
likewise

jarohen commented on crux-http-server/src/crux/http_server/entity.clj:
```diff
@@ -197,85 +224,65 @@
-         :entity-history (cio/fmap-cursor (fn [entity-history]
-                                            (map #(update % :crux.db/content-hash str) entity-history))
-                                          entity-history)}))
+      {:entity-history (cio/fmap-cursor identity entity-history)})
```
`(fmap-cursor identity entity-history)` === `entity-history`

jarohen commented on crux-http-server/src/crux/http_server/query.clj:
```diff
@@ -63,56 +102,44 @@
-    (cond-> {:find (c/read-edn-string-with-readers find)
-             :where (->> where vectorize-param (mapv c/read-edn-string-with-readers))
+
+;; TODO: Doesn't support all clauses a query can have - should do.
```
what we doing about this TODO? :smile:

jarohen commented on docs/reference/modules/ROOT/pages/http.adoc:
```diff
@@ -1,6 +1,6 @@
 = HTTP
 
-Crux offers a REST API layer in the `crux-http-server` module that allows you to send transactions and run queries over HTTP.
+Crux offers a REST API layer in the `crux-http-server` module that allows you to send transactions and run queries over HTTP. (See the REST API reference xref:http-api.adoc[here])
```
```
Crux offers a xref:...[REST API] layer in the `crux-http-server` module that allows you to send transactions and run queries over HTTP.
```
?

Also, think we talked about putting the REST API at the bottom of this file and linking to it?

jarohen replied:
Ah, we did, and you have - think this is a broken link now?

danmason replied:
It is indeed - will sort!

jarohen commented on docs/reference/modules/ROOT/pages/http.adoc:
```diff
@@ -73,183 +73,276 @@ If the node was started on `localhost:3000`, you can connect to it by doing the
-|<<#tx-log, `/tx-log`>>|GET| Returns a list of all transactions
-|<<#tx-log-post, `/tx-log`>>|POST|The "write" endpoint, to post transactions.
-|===
+All of the REST endpoints return both `application/edn` and `application/transit+json` (see transit https://github.com/cognitect/transit-clj[here]) - individual endpoints may return additional types - see the docs below. To set up your node to with the REST API, see the xref:http.adoc[HTTP Feature Module].
```
can use '`application/transit+json`' as the link text, rather than a separate 'see ...'

jarohen replied:
With adoc, I'm reliably informed (:smile:) that the idiom is one sentence per line

jarohen replied:
I'd break the sentence before 'individual endpoints', too.

Also, 'see the HTTP Feature module' - you're already there :)

jarohen replied:
in short:
```
All of the REST endpoints return both `application/edn` and xref:https://...[`application/transit+json`]
Individual endpoints may return additional types - see their docs below. 
```

jarohen commented on docs/reference/modules/ROOT/pages/http.adoc:
```diff
@@ -73,183 +73,276 @@ If the node was started on `localhost:3000`, you can connect to it by doing the
 
+.*Required Parameters*
+* `eid` (Crux ID)
+* `sort-order` (_required_, either `asc` or `desc`)
```
it's already in the 'required' section

jarohen commented on docs/reference/modules/ROOT/pages/http.adoc:
```diff
@@ -73,183 +73,276 @@ If the node was started on `localhost:3000`, you can connect to it by doing the
+[#query]
+=== GET `/_crux/query`
+
+Takes a datalog query and returns its results. Results are also available in `text/csv` and `text/tsv` formats (can force negotiation of these by using the `/_crux/query.csv` and `/_crux/query.tsv` endpoints respectively).
```
I wonder whether we should extend this both to `.edn`, and to the other endpoints, as a general pattern - wdyt?

(Maybe not one for this PR)

danmason replied:
Perhaps - though it does wind up making the handler pretty messy and you wind up repeating a lot of things - perhaps if there were a cleaner way of doing it?

danmason replied:
As discussed - not on this PR

jarohen commented on docs/reference/modules/ROOT/pages/http.adoc:
```diff
@@ -73,183 +73,276 @@ If the node was started on `localhost:3000`, you can connect to it by doing the
+==== Query Parameters
+
+.*Required Parameters*
+* `q` (URL Encoded datalog query)
```
lowercase e

jarohen reviewed (changes requested):
Lots of comments below, but fundamentally this is a great PR, cheers Dan - much cleaner and more consistent now, and much less ad-hoc parameter encoding/decoding code.

re `/_crux/entity`'s `eid` param - I find it weird that this is EDN-encoded, it seems like it should be subject to a `Content-Type` header or something. Particularly, for future JSON users, we'd be saying 'everything on the API can send/receive JSON, except this query param, which has to be EDN'. 

Not one for this PR, but requires some thought :thinking:

danmason committed:
Migrate both handlers to reitit - remove old routes, update client.

- Migrate original handler routes.
- Migrate data-browser routes + serve resources.
- Replace old query/entity endpoints
- fixup remote-api-client.
- Add spec + coercion to the console endpoints.
- `[crux.api :as api]` -> `[crux.api :as crux]`
- Add a default muuntaja instance, use muuntaja for request/response formatting.
- Refactor, clean and fixup http-server.entity
- Updates to behaviour of entity-history and query
- Fixes to tests and improvement of errors.
-- Handle coercion erors properly.
-- Throw errors (where appropriate) within entity/query on console
-- Errors are now handled/passed through muuntaja
-- Error handling improvement - using ce/illegal-arg
- All keys are parsed + set in query-params - optional params done
within `some->` - all nil keys are removed from actual query-params
before the request is sent to the server.
- Prefix all routes with `_crux/`
- POST `/_crux/tx-log` -> POST `/_crux/submit-tx`
- Add POST to query - take query as `query?=` only on GET

danmason committed:
Update REST documentation for HTTP API

jarohen committed:
Using standard reitit exception middleware

jarohen committed:
add 'cause' to crux.IAE
